File: studio/driver.py
"""Gated task driver — studio orchestration over harness primitives."""

import json
import os
import shutil

from harness import analysis, gitops, guard, packets, providers, supervisor

from studio import config, consent, draft, gates


def build_prompt_context(project):
    from studio import knowledge

    return knowledge.build_prompt_context(project)


class _DriverHooks(supervisor.Hooks):
    def __init__(self, jw):
        self.jw = jw

    def emit(self, event, **fields):
        self.jw.emit(event, **fields)

    def check(self):
        pass


def _planner_model(cfg, project):
    prov = (project.get("providers") or {}).get("planner")
    if prov and prov.get("model"):
        return prov["model"]
    return cfg["models"]["planner"]


def _stage_provider(cfg, project, stage, policy=None, run_id=None):
    merged = {**cfg, "providers": project.get("providers")}
    stage_cfg = (project.get("providers") or {}).get(stage) or {}
    kind = stage_cfg.get("kind", "ollama")
    model = stage_cfg.get("model", "")
    prov = providers.get(merged, stage, policy=policy)
    if run_id and kind in consent._CLOUD_KINDS:
        return consent.wrap_provider(prov, project, run_id, stage, kind, model)
    return prov


def _free_gb(path):
    return shutil.disk_usage(path).free // 2**30


def _task_packets_dir(cfg, task_id):
    return os.path.join(cfg["packets_dir"], task_id)


def _write_packet(cfg, task_id, name, body):
    pdir = _task_packets_dir(cfg, task_id)
    os.makedirs(pdir, exist_ok=True)
    packets.atomic_write_json(os.path.join(pdir, name), body)


def _terminal(jw, event, task_id, **fields):
    jw.emit(event, task=task_id, **fields)


def _incident(cfg, task_id, evidence):
    _write_packet(
        cfg, task_id, "incident.json", packets.packet("INCIDENT", task_id, **evidence)
    )


def _write_plan_md(cfg, task_id, task, plan, explain_fields, sim_fields):
    """Human-readable plan document next to the JSON packets."""
    blast = sim_fields.get("blast_files") or []
    lines = [
        f"# Plan — {task['title']}",
        "",
        f"Run: `{task_id}`",
        "",
        "## Approach",
        "",
        plan["approach"],
        "",
        "## Files",
        "",
        *[f"- `{p}`" for p in plan["files"]],
        "",
        "## Summary",
        "",
        explain_fields.get("summary", "(unavailable)"),
        "",
        "## Blast radius",
        "",
        f"Risk: {sim_fields.get('risk', '—')} · "
        f"{sim_fields.get('blast_count', 0)} files touched "
        f"({sim_fields.get('size_band', '—')})",
        "",
        *([f"- `{p}`" for p in blast] or ["- no direct importers detected"]),
        "",
    ]
    path = os.path.join(_task_packets_dir(cfg, task_id), "plan.md")
    with open(path, "w", encoding="utf-8") as f:
        f.write("\n".join(lines))


def _load_plan_packet(cfg, task_id):
    path = os.path.join(_task_packets_dir(cfg, task_id), "plan.json")
    with open(path, encoding="utf-8") as f:
        body = json.load(f)
    return {"approach": body["approach"], "files": body["files"]}


def _accept_evidence(cfg, task_id):
    wt = os.path.join(cfg["worktree_root"], task_id)
    branch = gitops.git(["rev-parse", "--abbrev-ref", "HEAD"], wt).strip()
    return {"worktree": wt, "branch": branch}


def _personalize_draft(cfg, project, run_dir, task, choice, jw, policy, run_id):
    """LLM-customize a copy of the chosen draft into draft-custom.html.

    Best-effort: reads the run-dir copy, never the template files, and any
    failure just leaves the un-personalized draft in place.
    """
    try:
        idx = int(choice.get("variant", 0))
        with open(os.path.join(run_dir, f"draft-{idx}.html"), encoding="utf-8") as f:
            base = f.read()
        prov = _stage_provider(cfg, project, "planner", policy=policy, run_id=run_id)
        custom = draft.customize(
            prov, _planner_model(cfg, project), base, task["prompt"]
        )
        if custom:
            draft.write_custom(run_dir, custom)
            jw.emit("draft_custom_ready")
    except Exception:
        pass


def run_task(
    project,
    run_id,
    task,
    cfg,
    implementer,
    gate_timeout_s=3600,
    resume_from=None,
    policy=None,
):
    """Execute one gated studio task; returns terminal status string."""
    run_dir = config.run_dir(project["id"], run_id)
    jw = packets.JournalWriter(cfg["journal_path"], run_id)
    hooks = _DriverHooks(jw)
    task_id = task["id"]
    pdir = _task_packets_dir(cfg, task_id)
    os.makedirs(pdir, exist_ok=True)
    os.makedirs(cfg["worktree_root"], exist_ok=True)
    gitops.ensure_integration(cfg["repo_root"], cfg["integration_branch"])

    ctx = build_prompt_context(project)
    work_task = dict(task)
    if ctx:
        work_task = {**work_task, "prompt": f"{ctx}\n\n{work_task['prompt']}"}

    if work_task.get("category") == "build":
        # greenfield builds legitimately touch many files; the default
        # max_plan_files cap is sized for fix/feature diffs and would
        # reject every new-site plan as "scope explosion"
        cfg = {
            **cfg,
            "limits": {
                **cfg["limits"],
                "max_plan_files": max(30, cfg["limits"]["max_plan_files"]),
            },
        }

    if resume_from == "accept":
        evidence = _accept_evidence(cfg, task_id)
        jw.emit("accept_ready")
        accept_gate = gates.wait(run_dir, "accept", gate_timeout_s, jw)
        if accept_gate.get("decision") == "approve":
            branch = evidence["branch"]
            wt = evidence["worktree"]
            gitops.remove_worktree(
                cfg["repo_root"], wt, task_id, keep_branch=True, branch_name=branch
            )
            hooks.emit("merge_start")
            supervisor.merge_shipped(cfg, work_task, branch_name=branch)
            gitops.git(["branch", "-D", branch], cfg["repo_root"])
            _terminal(jw, "shipped", task_id, evidence=evidence)
            return "SHIPPED"
        if accept_gate.get("decision") == "reject":
            gitops.remove_worktree(
                cfg["repo_root"],
                evidence["worktree"],
                task_id,
                keep_branch=False,
                branch_name=evidence["branch"],
            )
            _terminal(jw, "rejected_by_user", task_id, gate="accept")
            return "REJECTED"
        _terminal(jw, "paused", task_id, gate="accept")
        return "PAUSED"

    if resume_from != "plan":
        jw.emit("run_start", task=task_id)
        if work_task.get("category") == "build" and not work_task.get("edit_session"):
            # instant template draft variants; the user picks a design (or
            # skips) BEFORE any planning starts — a real gate, like plan/accept
            drafted = False
            try:
                tpl, variants = draft.make_variants(task["prompt"])
                draft.write_variants(run_dir, variants)
                draft.write_draft(run_dir, variants[0]["html"])
                jw.emit(
                    "draft_ready",
                    template=tpl,
                    variants=[{"id": v["id"], "label": v["label"]} for v in variants],
                )
                drafted = True
            except OSError:
                pass  # drafts are best-effort; skip the gate if they failed
            if drafted:
                design_gate = gates.wait(run_dir, "design", gate_timeout_s, jw)
                if design_gate.get("decision") == "timeout":
                    _terminal(jw, "paused", task_id, gate="design")
                    return "PAUSED"
                choice = design_gate.get("payload") or {}
                if choice.get("id"):
                    packets.atomic_write_json(
                        os.path.join(run_dir, "draft-choice.json"), choice
                    )
                    _personalize_draft(
                        cfg, project, run_dir, task, choice, jw, policy, run_id
                    )
        violations = guard.guard(work_task, cfg, _free_gb(cfg["repo_root"]))
        if violations:
            jw.emit("guard_reject", violations=violations)
            evidence = {"reason": "guard", "violations": violations}
            _incident(cfg, task_id, evidence)
            _terminal(jw, "escalated", task_id, evidence=evidence)
            return "ESCALATED"

        planner = _stage_provider(cfg, project, "planner", policy=policy, run_id=run_id)
        planner_model = _planner_model(cfg, project)
        hooks.emit("plan_start", model=planner_model)
        plan = supervisor.plan_task(planner, cfg, work_task)
        if plan is None:
            evidence = {"reason": "unplannable or scope explosion"}
            _incident(cfg, task_id, evidence)
            _terminal(jw, "escalated", task_id, evidence=evidence)
            return "ESCALATED"
        _write_packet(
            cfg, task_id, "plan.json", packets.packet("PLAN", task_id, **plan)
        )
        hooks.emit("plan_ready")

        jw.emit("explain_start")
        explain_body = analysis.explain(
            planner, planner_model, work_task, plan["files"], cfg["repo_root"]
        )
        explain_fields = {k: v for k, v in explain_body.items() if k != "kind"}
        _write_packet(
            cfg,
            task_id,
            "explain.json",
            packets.packet("EXPLAIN", task_id, **explain_fields),
        )
        jw.emit("explain_ready")

        sim_body = analysis.simulate(cfg["repo_root"], plan["files"])
        sim_fields = {k: v for k, v in sim_body.items() if k != "kind"}
        _write_packet(
            cfg,
            task_id,
            "simulate.json",
            packets.packet("SIMULATE", task_id, **sim_fields),
        )
        _write_plan_md(cfg, task_id, work_task, plan, explain_fields, sim_fields)
        jw.emit("simulate_ready")
    else:
        plan = _load_plan_packet(cfg, task_id)
        guard.guard(work_task, cfg, _free_gb(cfg["repo_root"]))

    plan_gate = gates.wait(run_dir, "plan", gate_timeout_s, jw)
    if plan_gate.get("decision") == "reject":
        _terminal(jw, "rejected_by_user", task_id, gate="plan")
        return "REJECTED"
    if plan_gate.get("decision") == "timeout":
        _terminal(jw, "paused", task_id, gate="plan")
        return "PAUSED"

    payload = plan_gate.get("payload") or {}
    approach = payload.get("approach", plan["approach"])
    files = payload.get("files", plan["files"])
    work_task = {
        **work_task,
        "plan_approach": approach,
        "plan_files": files,
    }

    working_file = work_task.get("working_file", "working.html")
    working_path = os.path.join(run_dir, working_file)
    if os.path.exists(working_path):
        work_task["prompt"] = (
            f"{work_task['prompt']}\n\n"
            f"Use {working_file} in the run directory as the approved static"
            " design source. Keep its sections and visual language unless"
            " the user plan explicitly asks for structural changes."
        )

    choice = draft.read_choice(run_dir)
    if choice:
        pack = next((p for p in draft.STYLE_PACKS if p["id"] == choice.get("id")), None)
        style = f"Visual direction: the user chose the '{choice.get('label')}' design draft."
        if pack and pack.get("accent"):
            style += (
                f" Use this palette — accent {pack['accent']},"
                f" text {pack['ink']}, background {pack['bg']},"
                f" surfaces {pack['surface']}; font stack {pack['font']}."
            )
        elif os.path.exists(os.path.join(run_dir, "draft-custom.html")):
            style += (
                " A personalized copy of that template is at"
                " draft-custom.html in the run directory; match its"
                " layout, sections, and content."
            )
        work_task["prompt"] = f"{work_task['prompt']}\n\n{style}"

    model = supervisor.pick_model(work_task, cfg)
    implementer_provider = _stage_provider(
        cfg, project, "implementer_llm", policy=policy, run_id=run_id
    )
    status, evidence = supervisor.attempt(
        work_task,
        model,
        cfg,
        implementer_provider,
        implementer,
        pdir,
        hooks=hooks,
        merge=False,
    )

    if status == "REVIEWED":
        jw.emit("accept_ready")
        accept_gate = gates.wait(run_dir, "accept", gate_timeout_s, jw)
        if accept_gate.get("decision") == "approve":
            branch = evidence["branch"]
            wt = evidence["worktree"]
            gitops.remove_worktree(
                cfg["repo_root"], wt, task_id, keep_branch=True, branch_name=branch
            )
            hooks.emit("merge_start")
            supervisor.merge_shipped(cfg, work_task, branch_name=branch)
            gitops.git(["branch", "-D", branch], cfg["repo_root"])
            _terminal(jw, "shipped", task_id, evidence=evidence)
            return "SHIPPED"
        if accept_gate.get("decision") == "reject":
            gitops.remove_worktree(
                cfg["repo_root"],
                evidence["worktree"],
                task_id,
                keep_branch=False,
                branch_name=evidence["branch"],
            )
            _terminal(jw, "rejected_by_user", task_id, gate="accept")
            return "REJECTED"
        _terminal(jw, "paused", task_id, gate="accept")
        return "PAUSED"

    _incident(cfg, task_id, evidence)
    event = status.lower()
    _terminal(jw, event, task_id, evidence=evidence)
    return status

File: src/server/driver.ts
/** Port of backend/studio/driver.py — gated task driver over harness primitives. */
import fs from "node:fs";
import path from "node:path";
import * as analysis from "./harness/analysis";
import * as gitops from "./harness/gitops";
import { guard } from "./harness/guard";
import { freeGb, mergeShipped, pickModel, planTask, attempt, type HarnessHooks, type ImplementerFn } from "./harness/supervisor";
import { buildPromptContext } from "./knowledge";
import { waitGate } from "./gates";
import { atomicWriteJson, JournalWriter, packet } from "./packets";
import { get, type Provider, type ProviderConfig } from "./providers";
import { wrapProvider } from "./consent";
import { readChoice } from "./runDraft";
import { STYLE_PACKS } from "./draft";
import type { LegacyProject } from "./legacyProjects";

const CLOUD_KINDS = new Set(["anthropic", "openai"]);

class DriverHooks implements HarnessHooks {
  constructor(private jw: JournalWriter) {}

  emit(event: string, fields: Record<string, unknown> = {}): void {
    this.jw.emit(event, fields);
  }

  check(): void {
    /* cooperative cancel checked via AbortSignal in waitGate/attempt loops */
  }
}

function plannerModel(cfg: Record<string, unknown>, project: LegacyProject): string {
  const prov = (project.providers as Record<string, { model?: string }> | null)?.planner;
  if (prov?.model) return prov.model;
  return (cfg.models as Record<string, string>).planner!;
}

function stageProvider(
  cfg: Record<string, unknown>,
  project: LegacyProject,
  stage: string,
  policy?: (stage: string, kind: string, model: string) => void,
  runId?: string,
): Provider {
  const merged: ProviderConfig = {
    ollamaUrl: String(cfg.ollama_url),
    providers: {
      ...((project.providers as Record<string, { kind?: string; model?: string; baseUrl?: string }>) ?? {}),
      ...((cfg.providers as Record<string, { kind?: string; model?: string; baseUrl?: string }>) ?? {}),
    },
  };
  const stageCfg = merged.providers?.[stage] ?? {};
  const kind = stageCfg.kind ?? "ollama";
  const model = stageCfg.model ?? "";
  const prov = get(merged, stage, policy);
  if (runId && CLOUD_KINDS.has(kind)) {
    return wrapProvider(prov, project, runId, stage, kind, model);
  }
  return prov;
}

function taskPacketsDir(cfg: Record<string, unknown>, taskId: string): string {
  return path.join(String(cfg.packets_dir), taskId);
}

function writePacket(cfg: Record<string, unknown>, taskId: string, name: string, body: unknown): void {
  const pdir = taskPacketsDir(cfg, taskId);
  fs.mkdirSync(pdir, { recursive: true });
  atomicWriteJson(path.join(pdir, name), body);
}

function terminal(jw: JournalWriter, event: string, taskId: string, fields: Record<string, unknown> = {}): void {
  jw.emit(event, { task: taskId, ...fields });
}

function incident(cfg: Record<string, unknown>, taskId: string, evidence: Record<string, unknown>): void {
  writePacket(cfg, taskId, "incident.json", packet("INCIDENT", taskId, evidence));
}

function writePlanMd(
  cfg: Record<string, unknown>,
  taskId: string,
  task: Record<string, unknown>,
  plan: { approach: string; files: string[] },
  explainFields: Record<string, unknown>,
  simFields: Record<string, unknown>,
): void {
  const blast = (simFields.blast_files as string[]) ?? [];
  const lines = [
    `# Plan — ${task.title}`,
    "",
    `Run: \`${taskId}\``,
    "",
    "## Approach",
    "",
    plan.approach,
    "",
    "## Files",
    "",
    ...plan.files.map((p) => `- \`${p}\``),
    "",
    "## Summary",
    "",
    String(explainFields.summary ?? "(unavailable)"),
    "",
    "## Blast radius",
    "",
    `Risk: ${simFields.risk ?? "—"} · ${simFields.blast_count ?? 0} files touched (${simFields.size_band ?? "—"})`,
    "",
    ...(blast.length ? blast.map((p) => `- \`${p}\``) : ["- no direct importers detected"]),
    "",
  ];
  fs.writeFileSync(path.join(taskPacketsDir(cfg, taskId), "plan.md"), lines.join("\n"), "utf8");
}

function loadPlanPacket(cfg: Record<string, unknown>, taskId: string): { approach: string; files: string[] } {
  const body = JSON.parse(fs.readFileSync(path.join(taskPacketsDir(cfg, taskId), "plan.json"), "utf8")) as Record<
    string,
    unknown
  >;
  return { approach: String(body.approach), files: body.files as string[] };
}

function acceptEvidence(cfg: Record<string, unknown>, taskId: string): { worktree: string; branch: string } {
  const wt = path.join(String(cfg.worktree_root), taskId);
  const branch = gitops.git(["rev-parse", "--abbrev-ref", "HEAD"], wt).trim();
  return { worktree: wt, branch };
}

export async function runTask(
  project: LegacyProject,
  runId: string,
  task: Record<string, unknown>,
  cfg: Record<string, unknown>,
  implementer: ImplementerFn,
  gateTimeoutS = 3600,
  resumeFrom: string | null = null,
  policy?: (stage: string, kind: string, model: string) => void,
  signal?: AbortSignal,
): Promise<string> {
  const runDir = path.dirname(String(cfg.journal_path));
  const jw = new JournalWriter(String(cfg.journal_path), runId);
  const hooks = new DriverHooks(jw);
  const taskId = String(task.id);
  const pdir = taskPacketsDir(cfg, taskId);
  fs.mkdirSync(pdir, { recursive: true });
  fs.mkdirSync(String(cfg.worktree_root), { recursive: true });
  gitops.ensureIntegration(String(cfg.repo_root), String(cfg.integration_branch));

  const ctx = buildPromptContext(project);
  let workTask = { ...task };
  if (ctx) workTask = { ...workTask, prompt: `${ctx}\n\n${workTask.prompt}` };

  let workCfg = cfg;
  if (workTask.category === "build") {
    const limits = cfg.limits as Record<string, number>;
    workCfg = {
      ...cfg,
      limits: { ...limits, max_plan_files: Math.max(30, limits.max_plan_files ?? 8) },
    };
  }

  if (resumeFrom === "accept") {
    const evidence = acceptEvidence(workCfg, taskId);
    jw.emit("accept_ready");
    const acceptGate = await waitGate(runDir, "accept", gateTimeoutS, jw, signal);
    if (acceptGate.decision === "approve") {
      gitops.removeWorktree(String(workCfg.repo_root), evidence.worktree, taskId, true, evidence.branch);
      hooks.emit("merge_start");
      await mergeShipped(workCfg, workTask, evidence.branch);
      gitops.git(["branch", "-D", evidence.branch], String(workCfg.repo_root));
      terminal(jw, "shipped", taskId, { evidence });
      return "SHIPPED";
    }
    if (acceptGate.decision === "reject") {
      gitops.removeWorktree(String(workCfg.repo_root), evidence.worktree, taskId, false, evidence.branch);
      terminal(jw, "rejected_by_user", taskId, { gate: "accept" });
      return "REJECTED";
    }
    terminal(jw, "paused", taskId, { gate: "accept" });
    return "PAUSED";
  }

  let plan: { approach: string; files: string[] };

  if (resumeFrom !== "plan") {
    jw.emit("run_start", { task: taskId });
    const violations = await guard(workTask, workCfg, freeGb(String(workCfg.repo_root)));
    if (violations.length) {
      jw.emit("guard_reject", { violations });
      incident(workCfg, taskId, { reason: "guard", violations });
      terminal(jw, "escalated", taskId, { evidence: { reason: "guard", violations } });
      return "ESCALATED";
    }

    const planner = stageProvider(workCfg, project, "planner", policy, runId);
    const pmodel = plannerModel(workCfg, project);
    hooks.emit("plan_start", { model: pmodel });
    const planned = await planTask(planner, workCfg, workTask);
    if (!planned) {
      incident(workCfg, taskId, { reason: "unplannable or scope explosion" });
      terminal(jw, "escalated", taskId, { evidence: { reason: "unplannable or scope explosion" } });
      return "ESCALATED";
    }
    plan = planned;
    writePacket(workCfg, taskId, "plan.json", packet("PLAN", taskId, plan));
    hooks.emit("plan_ready");

    jw.emit("explain_start");
    const explainBody = await analysis.explain(planner, pmodel, workTask, plan.files, String(workCfg.repo_root));
    const explainFields = Object.fromEntries(Object.entries(explainBody).filter(([k]) => k !== "kind"));
    writePacket(workCfg, taskId, "explain.json", packet("EXPLAIN", taskId, explainFields));
    jw.emit("explain_ready");

    const simBody = analysis.simulate(String(workCfg.repo_root), plan.files);
    const simFields = Object.fromEntries(Object.entries(simBody).filter(([k]) => k !== "kind"));
    writePacket(workCfg, taskId, "simulate.json", packet("SIMULATE", taskId, simFields));
    writePlanMd(workCfg, taskId, workTask, plan, explainFields, simFields);
    jw.emit("simulate_ready");
  } else {
    plan = loadPlanPacket(workCfg, taskId);
    await guard(workTask, workCfg, freeGb(String(workCfg.repo_root)));
  }

  const planGate = await waitGate(runDir, "plan", gateTimeoutS, jw, signal);
  if (planGate.decision === "reject") {
    terminal(jw, "rejected_by_user", taskId, { gate: "plan" });
    return "REJECTED";
  }
  if (planGate.decision === "timeout") {
    terminal(jw, "paused", taskId, { gate: "plan" });
    return "PAUSED";
  }

  const payload = (planGate.payload as Record<string, unknown>) ?? {};
  workTask = {
    ...workTask,
    plan_approach: payload.approach ?? plan.approach,
    plan_files: payload.files ?? plan.files,
  };

  const workingFile = String(workTask.working_file ?? "working.html");
  const workingPath = path.join(runDir, workingFile);
  if (fs.existsSync(workingPath)) {
    workTask.prompt = `${workTask.prompt}\n\nUse ${workingFile} in the run directory as the approved static design source. Keep its sections and visual language unless the user plan explicitly asks for structural changes.`;
  }

  const choice = readChoice(runDir);
  if (choice) {
    const pack = STYLE_PACKS.find((p) => p.id === choice.id);
    let style = `Visual direction: the user chose the '${choice.label}' design draft.`;
    if (pack?.accent) {
      style += ` Use this palette — accent ${pack.accent}, text ${pack.ink}, background ${pack.bg}, surfaces ${pack.surface}; font stack ${pack.font}.`;
    } else if (fs.existsSync(path.join(runDir, "draft-custom.html"))) {
      style += " A personalized copy of that template is at draft-custom.html in the run directory; match its layout, sections, and content.";
    }
    workTask.prompt = `${workTask.prompt}\n\n${style}`;
  }

  const model = pickModel(workTask, workCfg);
  const merged = {
    ...((workCfg.providers as Record<string, unknown>) ?? {}),
    ...((project.providers as Record<string, unknown>) ?? {}),
  };
  const reviewStage = merged.reviewer ? "reviewer" : "implementer_llm";
  const reviewerProvider = stageProvider(workCfg, project, reviewStage, policy, runId);
  const [status, evidence] = await attempt(
    workTask,
    model,
    workCfg,
    reviewerProvider,
    implementer,
    pdir,
    hooks,
    undefined,
    false,
  );

  if (status === "REVIEWED") {
    jw.emit("accept_ready");
    const acceptGate = await waitGate(runDir, "accept", gateTimeoutS, jw, signal);
    if (acceptGate.decision === "approve") {
      const ev = evidence as { branch: string; worktree: string };
      gitops.removeWorktree(String(workCfg.repo_root), ev.worktree, taskId, true, ev.branch);
      hooks.emit("merge_start");
      await mergeShipped(workCfg, workTask, ev.branch);
      gitops.git(["branch", "-D", ev.branch], String(workCfg.repo_root));
      terminal(jw, "shipped", taskId, { evidence });
      return "SHIPPED";
    }
    if (acceptGate.decision === "reject") {
      const ev = evidence as { branch: string; worktree: string };
      gitops.removeWorktree(String(workCfg.repo_root), ev.worktree, taskId, false, ev.branch);
      terminal(jw, "rejected_by_user", taskId, { gate: "accept" });
      return "REJECTED";
    }
    terminal(jw, "paused", taskId, { gate: "accept" });
    return "PAUSED";
  }

  incident(workCfg, taskId, evidence);
  terminal(jw, status.toLowerCase(), taskId, { evidence });
  return status;
}
